src/library/scala/concurrent/Future.scala

```diff
@@ -115,9 +115,8 @@ trait Future[+T] extends Awaitable[T] {
    */
   @deprecated("use `foreach` or `onComplete` instead (keep in mind that they take total rather than partial functions)", "2.12.0")
   def onSuccess[U](pf: PartialFunction[T, U])(implicit executor: ExecutionContext): Unit = onComplete {
-    case Success(v) =>
-      pf.applyOrElse[T, Any](v, Predef.identity[T]) // Exploiting the cached function to avoid MatchError
-    case _ =>
+    t =>
```

Comment:
Just out of curiosity: In which ways is this manual version superior to the pattern match?

Reply:
@sirthias Great question, let me illustrate an example:

~~~scala
class F {
  import scala.util._
  def _match(t: Try[Int]): Int =
    t match {
     case s @ Success(_) => s.value.hashCode
     case _ => 0
  }

  def _instanceOf(t: Try[Int]): Int =
    if (t.isInstanceOf[Success[Int]]) t.asInstanceOf[Success[Int]].value.hashCode
    else 0 
}
~~~

:javap F except:

~~~javap
  public int _match(scala.util.Try<java.lang.Object>);
    descriptor: (Lscala/util/Try;)I
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=5, args_size=2
         0: aload_1
         1: astore_3
         2: aload_3
         3: instanceof    #18                 // class scala/util/Success
         6: ifeq          27
         9: aload_3
        10: checkcast     #18                 // class scala/util/Success
        13: astore        4
        15: aload         4
        17: invokevirtual #22                 // Method scala/util/Success.value:()Ljava/lang/Object;
        20: invokevirtual #26                 // Method java/lang/Object.hashCode:()I
        23: istore_2
        24: goto          35
        27: goto          30
        30: iconst_0
        31: istore_2
        32: goto          35
        35: iload_2
        36: ireturn
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      37     0  this   L$line7/$read$$iw$$iw$F;
            0      37     1     t   Lscala/util/Try;
      LineNumberTable:
        line 14: 0
        line 15: 2
        line 16: 30
      StackMapTable: number_of_entries = 3
        frame_type = 253 /* append */
          offset_delta = 27
          locals = [ top, class scala/util/Try ]
        frame_type = 2 /* same */
        frame_type = 255 /* full_frame */
          offset_delta = 4
          locals = [ class $line7/$read$$iw$$iw$F, class scala/util/Try, int, class scala/util/Try ]
          stack = []
    Signature: #42                          // (Lscala/util/Try<Ljava/lang/Object;>;)I
    MethodParameters:
      Name                           Flags
      t                              final

  public int _instanceOf(scala.util.Try<java.lang.Object>);
    descriptor: (Lscala/util/Try;)I
    flags: ACC_PUBLIC
    Code:
      stack=1, locals=2, args_size=2
         0: aload_1
         1: instanceof    #18                 // class scala/util/Success
         4: ifeq          20
         7: aload_1
         8: checkcast     #18                 // class scala/util/Success
        11: invokevirtual #22                 // Method scala/util/Success.value:()Ljava/lang/Object;
        14: invokevirtual #26                 // Method java/lang/Object.hashCode:()I
        17: goto          21
        20: iconst_0
        21: ireturn
      LocalVariableTable:
        Start  Length  Slot  Name   Signature
            0      22     0  this   L$line7/$read$$iw$$iw$F;
            0      22     1     t   Lscala/util/Try;
      LineNumberTable:
        line 20: 0
        line 21: 20
      StackMapTable: number_of_entries = 2
        frame_type = 20 /* same */
        frame_type = 64 /* same_locals_1_stack_item */
          stack = [ int ]
    Signature: #42                          // (Lscala/util/Try<Ljava/lang/Object;>;)I
    MethodParameters:
      Name                           Flags
      t                              final
~~~

Reply:
In other words, patmat very quickly exhausts the inlining budget.

Reply:
Ok, I see.
The code that'll be actually running (post-JIT) is practically identical, but due to scalac producing less compact code for the pattern match the inlining thresholds, which are based on (pre-JIT) bytecode instruction counts, favor the manual version.
Thanks!

Reply:
Yes. That's my line of thinking :)